syntax = "proto3";

package market_plant.v1;

import "google/protobuf/empty.proto";

// Order Book Management

enum Side {
    SIDE_UNSPECIFIED = 0;
    BID = 1;
    ASK = 2;
}

message Level {
    Side side = 1;
    uint32 price = 2;
    uint32 quantity = 3;
}

enum OrderBookEventType {
    EVENT_UNSPECIFIED = 0;
    ADD_LEVEL = 1;
    REPLACE_LEVEL = 2;
    REMOVE_LEVEL = 3;
}

message OrderBookEventUpdate {
    OrderBookEventType type = 1;
    Level level = 2;
}

message SnapshotUpdate {
    repeated OrderBookEventUpdate bids = 1;
    repeated OrderBookEventUpdate asks = 2;
}

message IncrementalUpdate {
    OrderBookEventUpdate update = 1;
}

message OrderBookUpdate {
    uint32 instrument_id = 1;
    oneof update_type {
        SnapshotUpdate snapshot = 2;
        IncrementalUpdate incremental = 3;
    }
}

// Subscription Management

message InstrumentIds {
    repeated uint32 ids = 1;
}

message Subscription {
    oneof action {
        InstrumentIds subscribe = 1;
        InstrumentIds unsubscribe = 2;  
    }
}

message SubscriberInitialization {
    uint32 subscriber_id = 1;
    bytes session_id = 2;
}

message StreamResponse {
    oneof payload {
        SubscriberInitialization init = 1;
        OrderBookUpdate update = 2;
    }
}

message UpdateSubscriptionRequest {
    uint32 subscriber_id = 1;
    bytes session_id = 2;
    Subscription change = 3;
}

service MarketPlantService {
    // Server-side streaming
    rpc StreamUpdates ( Subscription ) returns (stream StreamResponse);

    // Control-plane for modifying subscriptions
    rpc UpdateSubscriptions ( UpdateSubscriptionRequest ) returns (google.protobuf.Empty);
}